system/src/ledger/ledger_manager.cpp

```diff
@@ -0,0 +1,1599 @@
+        if (contexts_.size() >= (int)MAX_LEDGER_COUNT) {
+            LOG(ERROR, "Maximum number of ledgers reached");
+            return SYSTEM_ERROR_LEDGER_TOO_MANY;
+        }
+        // Create a new sync context
+        newCtx.reset(new(std::nothrow) LedgerSyncContext());
+        if (!newCtx) {
+            return SYSTEM_ERROR_NO_MEMORY;
+        }
+        size_t n = strlcpy(newCtx->name, name, sizeof(newCtx->name));
+        if (n >= sizeof(newCtx->name)) {
+            return SYSTEM_ERROR_INVALID_ARGUMENT;
+        }
+        ctx = newCtx.get();
+    } else {
+        ctx = it->get();
+    }
+    // Create a ledger instance
+    auto lr = makeRefCountPtr<Ledger>(ctx);
+    if (!lr) {
+        return SYSTEM_ERROR_NO_MEMORY;
+    }
+    int r = lr->init(ctx->name);
+    if (r < 0) {
+        LOG(ERROR, "Failed to initialize ledger: %d", r);
+        return r;
+    }
+    if (!found) {
+        it = contexts_.insert(it, std::move(newCtx));
+        if (it == contexts_.end()) {
+            return SYSTEM_ERROR_NO_MEMORY;
+        }
+        if (state_ >= State::READY) {
+            // Request the info about the newly created ledger
+            setPendingState(ctx, PendingState::GET_INFO);
+            runNext();
+        }
+    }
+    ctx->instance = lr.get();
+    ledger = std::move(lr);
+    return 0;
+}
+
+int LedgerManager::getLedgerNames(Vector<CString>& namesArg) {
+    std::lock_guard lock(mutex_);
+    if (state_ == State::NEW) {
+        return SYSTEM_ERROR_INVALID_STATE;
+    }
+    // Return all ledgers found in the filesystem, not just the usable ones
+    FsLock fs;
+    Vector<CString> names;
+    lfs_dir_t dir = {};
+    int r = lfs_dir_open(fs.instance(), &dir, LEDGER_ROOT_DIR);
+    if (r == 0) {
+        SCOPE_GUARD({
+            int r = closeDir(fs.instance(), &dir);
+            if (r < 0) {
+                LOG(ERROR, "Failed to close directory handle: %d", r);
+            }
+        });
+        lfs_info entry = {};
+        while ((r = lfs_dir_read(fs.instance(), &dir, &entry)) == 1) {
+            if (entry.type != LFS_TYPE_DIR) {
+                LOG(WARN, "Found unexpected entry in ledger directory");
+                continue;
+            }
+            if (std::strcmp(entry.name, ".") == 0 || std::strcmp(entry.name, "..") == 0) {
+                continue;
+            }
+            CString name(entry.name);
+            if (!name || !names.append(std::move(name))) {
+                return SYSTEM_ERROR_NO_MEMORY;
+            }
+        }
+        CHECK_FS(r);
+    } else if (r != LFS_ERR_NOENT) {
+        CHECK_FS(r); // Forward the error
+    }
+    namesArg = std::move(names);
+    return 0;
+}
+
+int LedgerManager::removeLedgerData(const char* name) {
+    if (!*name) {
+        return SYSTEM_ERROR_INVALID_ARGUMENT;
+    }
+    std::lock_guard lock(mutex_);
+    // TODO: Allow removing ledgers regardless of the device state or if the given ledger is in use
+    if (state_ != State::OFFLINE) {
```

Comment:
This will be important to document:
To remove the Ledger data from a device (for example, during development or to simulate the first cloud connection), go offline `Particle.disconnect()` then purge the Ledgers `Ledger::removeAll()`

Correct this if it's wrong.

Reply:
Just disconnecting from the cloud won't be enough if the application has valid references to the ledger instances being deleted. The most reliable way would probably be to delete ledgers early on startup, before a cloud connection is initiated and any ledger instances are created.

This is quite a limitation so I think removing it should be one of the first improvements we do the ledger API once this PR is merged.

Reply:
It doesn't sound too bad in practice as removing a Ledger from a device isn't a normal thing that a program has to do.
Did you end up adding a Wiring API to declare which Ledgers are used by an application, so the cleanup of unused Ledgers is automatic?

Reply:
Yep, there's `Particle.useLedgers()` method that does that: https://github.com/particle-iot/device-os/blob/07308ada47f7344250e7a06a281ca3d006a5f7f4/wiring/inc/spark_wiring_cloud.h#L471-L478.